src/codegen/wasm.zig

```diff
@@ -747,94 +754,59 @@ pub const Context = struct {
-        };
-
+        const opcode: wasm.Opcode = buildOpcode(.{
+            .op = op,
+            .valtype1 = try self.typeToValtype(inst.base.src, inst.base.ty),
+        });
         try self.code.append(wasm.opcode(opcode));
         return .none;
     }
 
     fn emitConstant(self: *Context, inst: *Inst.Constant) InnerError!void {
         const writer = self.code.writer();
-        switch (inst.base.ty.tag()) {
-            .u32 => {
-                try writer.writeByte(wasm.opcode(.i32_const));
-                try leb.writeILEB128(writer, inst.val.toUnsignedInt());
+        switch (inst.base.ty.zigTypeTag()) {
+            .Int => {
+                // write opcode
+                const opcode: wasm.Opcode = buildOpcode(.{
+                    .op = .@"const",
+                    .valtype1 = try self.typeToValtype(inst.base.src, inst.base.ty),
+                });
+                try writer.writeByte(wasm.opcode(opcode));
+                // write constant
+                switch (inst.base.ty.intInfo(self.target).signedness) {
+                    .signed => try leb.writeILEB128(writer, inst.val.toSignedInt()),
+                    .unsigned => try leb.writeILEB128(writer, inst.val.toUnsignedInt()),
+                }
             },
-            .i32, .bool => {
+            .Bool => {
+                // write opcode
                 try writer.writeByte(wasm.opcode(.i32_const));
+                // write constant
                 try leb.writeILEB128(writer, inst.val.toSignedInt());
             },
-            .u64 => {
-                try writer.writeByte(wasm.opcode(.i64_const));
-                try leb.writeILEB128(writer, inst.val.toUnsignedInt());
-            },
-            .i64 => {
-                try writer.writeByte(wasm.opcode(.i64_const));
-                try leb.writeILEB128(writer, inst.val.toSignedInt());
-            },
-            .f32 => {
-                try writer.writeByte(wasm.opcode(.f32_const));
-                // TODO: enforce LE byte order
-                try writer.writeAll(mem.asBytes(&inst.val.toFloat(f32)));
-            },
-            .f64 => {
-                try writer.writeByte(wasm.opcode(.f64_const));
-                // TODO: enforce LE byte order
-                try writer.writeAll(mem.asBytes(&inst.val.toFloat(f64)));
+            .Float => {
+                // write opcode
+                const opcode: wasm.Opcode = buildOpcode(.{
+                    .op = .@"const",
+                    .valtype1 = try self.typeToValtype(inst.base.src, inst.base.ty),
+                });
+                try writer.writeByte(wasm.opcode(opcode));
+                // write constant
+                switch (inst.base.ty.floatBits(self.target)) {
+                    0...32 => try writer.writeIntLittle(u32, @bitCast(u32, inst.val.toFloat(f32))),
+                    64 => try writer.writeIntLittle(u64, @bitCast(u64, inst.val.toFloat(f64))),
```

Comment:
I had a shot at fixing
```TODO: enforce LE byte order```
and this is what I found, but not sure if it's correct. Could you help me check?

Reply:
Looks great!